packages/jest-mock/src/__tests__/whenCalledWith.test.ts

```diff
@@ -0,0 +1,282 @@
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('special output');
+  });
+
+  it('supports matchers', () => {
+    const fn = moduleMocker.fn();
+    fn.mockReturnValue('hello');
+    fn.whenCalledWith(expect.any(String)).mockImplementation(
+      v => `hello, ${v}`,
+    );
+    fn.whenCalledWith('hello', expect.any(String)).mockImplementation(
+      (_, y) => `morning, ${y}`,
+    );
+
+    expect(fn(3)).toBe('hello');
+    expect(fn('world')).toBe('hello, world');
+    expect(fn('hello', 3)).toBe('hello');
+    expect(fn('hello', 'friend')).toBe('morning, friend');
+  });
+
+  it('still tracks all calls on the root mock', () => {
+    const fn = moduleMocker.fn();
+    fn.mockReturnValue('generic output');
+    fn.whenCalledWith('special input').mockReturnValue('special output');
+
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('special output');
+    expect(fn).toHaveBeenNthCalledWith(1, 'arbitrary');
+    expect(fn).toHaveBeenNthCalledWith(2, 'special input');
+    expect(fn).toHaveBeenCalledTimes(2);
+  });
+
+  it('supports spies', () => {
+    class TestClass {
+      public calls: Record<string, number> = {};
+      increment(arg: string) {
+        this.calls[arg] = (this.calls[arg] ?? 0) + 1;
+        return this.calls[arg];
+      }
+    }
+
+    const instance = new TestClass();
+    const spy = spyOn(instance, 'increment');
+    spy.whenCalledWith('hello').mockReturnValue(-3);
+
+    expect(instance.increment('goodbye')).toBe(1);
+    expect(instance.increment('goodbye')).toBe(2);
+    expect(instance.increment('hello')).toBe(-3);
+    expect(instance.increment('hello')).toBe(-3);
+    expect(instance.increment('goodbye')).toBe(3);
+  });
+
+  it('gets removed by reset', () => {
+    const fn = moduleMocker.fn();
+    fn.mockReturnValue('generic output');
+    fn.whenCalledWith('special input').mockReturnValue('special output');
+
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('special output');
+
+    fn.mockReset();
+    expect(fn('arbitrary')).toBeUndefined();
+    expect(fn('special input')).toBeUndefined();
+  });
+
+  it('gets removed by resetAllMocks', () => {
+    const fn = moduleMocker.fn();
+    fn.mockReturnValue('generic output');
+    fn.whenCalledWith('special input').mockReturnValue('special output');
+
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('special output');
+
+    moduleMocker.resetAllMocks();
+    expect(fn('arbitrary')).toBeUndefined();
+    expect(fn('special input')).toBeUndefined();
+  });
+
+  it('gets clobbered by other mocks', () => {
+    const fn = moduleMocker.fn();
+    fn.whenCalledWith('special input').mockReturnValue('special output');
+    fn.mockReturnValue('generic output');
+
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('generic output');
+  });
```

Comment:
Hmm I'm not sure I agree with this behavior. I guess it makes sense it would clobber.

Reply:
Yeah, I know some libraries do this where a `when` takes precedence over a regular mock. But then you also want `when('special', 3)` to take precedence over `when(Any(), 3)`, and with arbitrary asymmetric matchers that gets impossible very fast. (Not to mention: how does that interact with `mock*Once`/`withImplementation`?) So instead we just give later mocks precedence, such that the behavior is clear.